server/src/main/java/org/elasticsearch/cluster/coordination/MasterHistory.java

```diff
@@ -0,0 +1,214 @@
+     */
+    public boolean hasSeenMasterInLastNSeconds(int n) {
+        List<TimeAndMaster> masterHistoryCopy = getMasterHistoryForLast30Minutes();
+        long now = nowSupplier.get();
+        TimeValue nSeconds = new TimeValue(n, TimeUnit.SECONDS);
+        long nSecondsAgo = now - nSeconds.getMillis();
+        return getCurrentMaster() != null
+            || masterHistoryCopy.stream().anyMatch(timeAndMaster -> timeAndMaster.time > nSecondsAgo && timeAndMaster.master != null);
+    }
+
+    /*
+     * This method creates a copy of masterHistory that only has entries from more than 30 minutes before now (but leaves the newest
+     * entry in even if it is more than 30 minutes old).
+     */
+    private List<TimeAndMaster> getMasterHistoryForLast30Minutes() {
+        List<TimeAndMaster> masterHistoryCopy = new ArrayList<>(masterHistory);
+        if (masterHistoryCopy.size() < 2) {
+            return masterHistoryCopy;
+        }
+        long now = nowSupplier.get();
+        TimeValue thirtyMinutes = new TimeValue(30, TimeUnit.MINUTES);
+        long thirtyMinutesAgo = now - thirtyMinutes.getMillis();
+        TimeAndMaster mostRecent = masterHistoryCopy.isEmpty() ? null : masterHistoryCopy.get(masterHistoryCopy.size() - 1);
+        masterHistoryCopy = masterHistoryCopy.stream()
+            .filter(timeAndMaster -> timeAndMaster.time > thirtyMinutesAgo)
+            .collect(Collectors.toList());
+        if (masterHistoryCopy.isEmpty() && mostRecent != null) { // The most recent entry was more than 30 minutes ago
+            masterHistoryCopy.add(mostRecent);
+        }
+        return masterHistoryCopy;
+    }
+
+    /**
+     * Clears out anything from masterHistory that is from more than 30 minutes before now (but leaves the newest entry in even if it is
+     * more than 30 minutes old). Rather than being scheduled, this method is called whenever the cluster state changes.
+     */
+    private void removeOldMasterHistory(List<TimeAndMaster> newMasterHistory) {
+        if (newMasterHistory.size() < 2) {
+            return;
+        }
+        long now = nowSupplier.get();
+        TimeValue thirtyMinutes = new TimeValue(30, TimeUnit.MINUTES);
+        long thirtyMinutesAgo = now - thirtyMinutes.getMillis();
+        TimeAndMaster mostRecent = newMasterHistory.isEmpty() ? null : newMasterHistory.get(newMasterHistory.size() - 1);
+        newMasterHistory.removeIf(timeAndMaster -> timeAndMaster.time < thirtyMinutesAgo);
+        if (newMasterHistory.isEmpty() && mostRecent != null) { // The most recent entry was more than 30 minutes ago
+            newMasterHistory.add(mostRecent);
+        }
+    }
+
+    /**
+     * This method returns an immutable view of this master history, typically for sending over the wire to another node.
+     * @return An immutable view of this master history
+     */
+    public List<DiscoveryNode> getImmutableView() {
```

Comment:
We don't want the times though (since this is used for remote machines), so regardless we've got to transform the list.

Reply:
I went with `getNodes()`. I'm not crazy about it but I admit it's better than `getImmutableView()`. I was originally trying to get across that this is a version that was not automatically updated with cluster state changes (this used to be a MasterHistory object rather than a List). Immutable is the wrong word -- more like "NotChangedFromUnderneath". Anyway, done.